### ISSUE TITLE
Memory Corruption / Double Free in grpc python client after forking

### ISSUE THREAD

**Issue by andrew-cl** [grpc/grpc#10140]
We're running a tornado web server that forks X processes and starts serving requests. We recently added a GRPC client (1.1.3) to talk to some new services, and have been getting memory corruption / double frees.

Example stderr msg:

*** Error in `/usr/bin/python': double free or corruption (!prev): 0x0000000001933f20 ***

Curiously, in the lifetime of the server(s), it's the same memory address that is getting double freed/corrupted. Left running long enough, the server will start to fail on requests because the in-memory python code files have been corrupted (i.e. classes missing class level variables).

Ex.
Class A(object):
    _class_var = 'ASDF'

print A._class_var

would raise an exception saying _class_var is missing from the class A.

We are unable to reproduce when running a single process, it only happens when we fork. Seems like there's a race going on in the C code somewhere, but haven't been able to track it down with strace/gdb...

We're running on python 2.7.6, ubuntu 14.04.

**Comment by nathanielmanistaatgoogle** [comment]
Forking isn't yet supported in gRPC Python; [this message](https://github.com/grpc/grpc/issues/7951#issuecomment-260811550) on [this issue](https://github.com/grpc/grpc/issues/7951) gets into some of the reasons why.

We're not sure when fork support will happen, if ever. How much pain is this causing you? How unpleasant is working around the inability to fork?

**Comment by kahuang** [comment]
To clarify, we don't instantiate any grpc stuff prefork, so post fork we're able to use grpc.

The issue is, after extended use with greenlets we'll get a memory corruption. We think it's some issue related to greenlets, and creating new threads in greenlets since we were able to reproduce without grpc.

That being said - the fork issue has been a problem for things that instantiate grpc connections prefork. We've worked around it for now by having a "grpc proxy" process that we talk to over http that does the grpc call for us, then returns the results. This is pretty messy, but works... Downsides include not being able to use streaming rpcs through the proxy since we're using http1 and serialization/deserialization overhead (which can add up to quite a bit).

Ideally this could work post fork, since its a bit of a headache to get the proxy process set up correctly everywhere + the perf impacts of having a proxy layer.
